[FIRRTL] Add utils to InnerSymAttr. NFC.

prithayan opened the pull request:
Add utils to `InnerSymAttr`  required for adding support to ports.
Splits changes from https://github.com/llvm/circt/pull/3582

uenoku commented on include/circt/Dialect/FIRRTL/FIRRTLAttributes.td:
```diff
@@ -232,20 +232,28 @@ def InnerSymAttr : AttrDef<FIRRTLDialect, "InnerSym"> {
   let builders = [
     AttrBuilderWithInferredContext<(ins "::mlir::StringAttr":$sym),[{
-      return get(sym.getContext(), 
+      if (!sym)
+        return InnerSymAttr();
```
Can we remove the gate `if(!sym) return ..`? I think it is users' responsibility to guarantee that the given attribute is non-null within `AttrBuilderWithInferredContext`, e.g https://github.com/llvm/llvm-project/blob/26dcc7e64afe089c78884213ee7f56723c95cb80/mlir/include/mlir/IR/BuiltinAttributes.td#L387-L392

prithayan replied:
Thanks @uenoku for this suggestion. Actually I think its better to remove this check. I updated the users to not rely on this check.

uenoku commented on lib/Dialect/FIRRTL/FIRRTLAttributes.cpp:
```diff
@@ -131,6 +145,14 @@ bool InnerSymAttr::all_of_props(
 
 size_t InnerSymAttr::numSymbols() { return getImpl()->props.size(); }
 
+bool InnerSymAttr::isSymbolInvalid() {
+  return all_of_props([](InnerSymPropertiesAttr p) {
+    if (p.getName().getValue().empty())
+      return true;
+    return false;
+  });
+}
+
```
Isn't is something we want to have in `InnerSymAttr::verify`?

dtzSiFive commented:
cc #3582 for comments on these, sorry didn't see this was split off or would've added here.  Most important is the mentioned ArrayRef-to-stack-vector issue.

prithayan commented:
> cc #3582 for comments on these, sorry didn't see this was split off or would've added here. Most important is the mentioned ArrayRef-to-stack-vector issue.

Thanks for your review on the other PR, and sorry I missed to link it. Addressed them and fixed this PR.

youngar commented:
Some random thing glancing at this PR, sorry if they have been brought up or addressed before.  Nothing blocking but maybe some ideas for the future:

1. What about adding a `begin()` and `end()` so you can iterate the prop array directly
2. `getAllSymNames` can be implemented as a mapped iterator instead of returning a SmallVector to be more efficient
3. might make more sense to call `getAllSymNames` as `getSymNames`, i.e. isn't "all" implied?
4. Not sure that `all_of_props` deserves to be a public API, are there other users of this?
5. I don't see any users of `hasSymNamed()` in this PR or #3582, are there other users of this?
6. Should this be called `isSymbolInvalid` or `isSymbolValid`?
```
    /// Check if symbol name is valid.
    bool isSymbolInvalid() const;
```
7. What about using an empty array to indicate an "invalid" `InnerSymAttr`, instead of one where all the strings in the array are empty?
8. Might make sense to have a `InnerSymAttr::getInvalid` or `getEmpty()` function to create an instance of the invalid inner sym attr.
9. In general it seems confusing calling this a `InnerSymAttr`, when conceptually it is a list of symbols.  E.g. the interface is split between functions that make it seem like there is only 1 symbol (`isSymbolValid`, `getSymName`), and functions that show there is actually more (`numSymbols`)

prithayan commented:
Thanks @youngar for your excellent suggestions. I am going to add the corresponding iterators.

> Some random thing glancing at this PR, sorry if they have been brought up or addressed before. Nothing blocking but maybe some ideas for the future:
> 
>     1. What about adding a `begin()` and `end()` so you can iterate the prop array directly
> 
>     2. `getAllSymNames` can be implemented as a mapped iterator instead of returning a SmallVector to be more efficient
> 
>     3. might make more sense to call `getAllSymNames` as `getSymNames`, i.e. isn't "all" implied?
> 

That makes sense, we can drop the "all".

>     4. Not sure that `all_of_props` deserves to be a public API, are there other users of this?
> 
You are right, `all_of_props` doesn't need to be public API.

>     5. I don't see any users of `hasSymNamed()` in this PR or [[FIRRTL] Update port symbols to InnerSymAttr #3582](https://github.com/llvm/circt/pull/3582), are there other users of this?
> 
It doesn't have a user after we recently dropped a function using it. We can remove it.

>     6. Should this be called `isSymbolInvalid` or `isSymbolValid`? The comment says it checks if its valid.  Maybe `isInvalid()`?
> 
> 
> ```
>     /// Check if symbol name is valid.
>     bool isSymbolInvalid() const;
> ```
> 
Actually the comment needs to be updated. 

>     7. What about using an empty array to indicate an "invalid" `InnerSymAttr`, instead of checking if all the strings in the array are empty?
> 
The problem was that, empty string is being used to indicate no symbol on ports. So, the `isSymbolInvalid` just checks if the symbol name is empty string to indicate the symbol does not exist (on ports). Now, we cannot just check the existence of empty string on `fieldID=0`, we have to do it for all `fieldID`, before inferring that the symbol name is invalid.

>     8. Might make sense to have a `InnerSymAttr::getInvalid` or `getEmpty()` (or just `get(MLIRContext)`) function to create an instance of the invalid inner sym attr.
> 
I like this idea of empty array, instead of empty string. Will do that.

>     9. In general it seems confusing calling this a `InnerSymAttr`, when conceptually it is a list of symbols.  E.g. the interface is split between functions that make it seem like there is only 1 symbol (`isSymbolValid`, `getSymName`), and functions that show there is actually more (`numSymbols`)
True, we need to do some brainstorming to better name this. This thought occurred to us several times earlier.

dtzSiFive commented on include/circt/Dialect/FIRRTL/InnerSymbolTable.h:
```diff
@@ -182,6 +182,30 @@ class InnerSymbolTable {
   static LogicalResult walkSymbols(Operation *op, InnerSymCallbackFn callback);
 
+  /// Get the sym name from the InnerSymPropertiesAttr.
+  static StringAttr getInnerSymName(const InnerSymPropertiesAttr p) {
```
If this method stays here, please align this "get inner symbol from a thing" method with the ones above `walkSymbols`, currently called `getInnerSymbol`.  Not adamant about which name but having both is confusing.

That said, hopefully users don't have to deal with `InnerSymProperties` too often?  They're only available via the new ISA iterators (begin/end) and have their own getName method so I'm not sure what this is useful for.

prithayan replied:
I agree this seemed a bit off. I just thought it would be nice to have, but there's no use case. So, got rid of it.

dtzSiFive commented on include/circt/Dialect/FIRRTL/InnerSymbolTable.h:
```diff
@@ -182,6 +182,30 @@ class InnerSymbolTable {
+  }
+
+  /// Get the iterator for all the symbol names for an InnerSymAttr.
+  static auto getSymNamesBegin(InnerSymAttr isa) {
```
This doesn't seem like the place for these methods, but I'm not sure where they should live.  Member functions on the attribute seems ideal, if tablegen doesn't make that too painful.

dtzSiFive commented on lib/Dialect/FIRRTL/FIRRTLAttributes.cpp:
```diff
@@ -122,14 +122,23 @@ StringAttr InnerSymAttr::getSymIfExists(unsigned fieldId) {
+
+bool InnerSymAttr::isSymbolInvalid() const { return getImpl()->props.empty(); }
+
+void InnerSymAttr::forAllSymNames(
```
Should this method return bool to indicate whether a failure was encountered?

Consider LogicalResult instead of bool, which can be unclear as to which way is success (and harder to ignore by accident).  Not always better though :+1: .

uenoku replied:
Do we need this helper function? `getSymNames()` looks sufficient to me
`llvm::all_of(inneSymAttr.getSymNames(), []())`

prithayan replied:
Renamed it and got rid of the `getSymNames` which was just copying around the array.

uenoku commented on include/circt/Dialect/FIRRTL/FIRRTLAttributes.td:
```diff
@@ -232,20 +232,34 @@ def InnerSymAttr : AttrDef<FIRRTLDialect, "InnerSym"> {
-    size_t numSymbols() ;
+    size_t numSymbols() const;
+    /// Check if this is an empty array, no sym names stored.
+    bool isSymbolInvalid() const;
```
+1 for renaming this to `empty/size`. It looks very clear to me

prithayan replied:
Thanks for the better name suggestions, updated and inlined them.

prithayan commented on include/circt/Dialect/FIRRTL/FIRRTLAttributes.td:
```diff
@@ -232,20 +232,34 @@ def InnerSymAttr : AttrDef<FIRRTLDialect, "InnerSym"> {
+    }]>,
+    // Create an empty array, represents an invalid InnerSym.
+    AttrBuilder<(ins),[{
+      return get($_ctxt, {});
```
True, right now only ports have to use this, just because of the way port symbols are stored. We should update the port symbols storage, so that we don't need invalid `InnerSymAttr` for ports with no symbols. Ideally we shouldn't have to rely on this for anything.

prithayan commented on include/circt/Dialect/FIRRTL/FIRRTLAttributes.td:
```diff
@@ -232,20 +232,34 @@ def InnerSymAttr : AttrDef<FIRRTLDialect, "InnerSym"> {
+    StringAttr getSymName() const;
     /// Check if all the elements of props satisfy the predicate lambda.
-    bool all_of_props(std::function<bool (InnerSymPropertiesAttr)>) ;
+    bool all_of_props(std::function<bool (InnerSymPropertiesAttr)>) const;
```
Initially seemed like a nice utility to avoid copying around the props array, but removed it and just using the array.

prithayan commented on include/circt/Dialect/FIRRTL/InnerSymbolTable.h:
```diff
@@ -182,6 +182,30 @@ class InnerSymbolTable {
+    InnerSymAttr isa =
+        op->getAttrOfType<InnerSymAttr>(getInnerSymbolAttrName());
+    if (isa)
+      llvm::make_range(getSymNamesBegin(isa), getSymNamesEnd(isa));
```
I think these iterator might not be required, got rid of them.

prithayan commented on include/circt/Dialect/FIRRTL/FIRRTLAttributes.td:
```diff
@@ -232,20 +232,34 @@ def InnerSymAttr : AttrDef<FIRRTLDialect, "InnerSym"> {
+    iterator end() const {return getProps().end();}
+    /// Invoke the func, for all sym names. Stop the iteration if the func
+    /// returns false, else continue.
+    void forAllSymNames(llvm::function_ref<bool (StringAttr)>) const;
```
Updated the name to make it `walkSymbols`, and right now this is required for recording all the symbol names by `ModuleNamespace` .

prithayan reviewed (commented):
Thanks for the detailed review and discussion on the naming. Updated the names and removed some unnecessary utilities.
